src/main/java/com/iexec/worker/executor/TaskExecutorService.java

```diff
@@ -212,24 +202,85 @@ private String compute(ContributionAuthorization contributionAuth) {
 
         if (isDatasetDecryptionNeeded && !isDatasetDecrypted) {
             customFeignClient.updateReplicateStatus(chainTaskId, COMPUTE_FAILED);
-            stdout = "Failed to decrypt dataset, URI:" + taskDescription.getDatasetUri();
-            log.error(stdout + " [chainTaskId:{}]", chainTaskId);
-            return stdout;
+            message = "Failed to decrypt dataset, URI:" + taskDescription.getDatasetUri();
+            log.error(message + " [chainTaskId:{}]", chainTaskId);
+            return message;
+        }
+
+        String error = checkContributionAbility(chainTaskId);
+        if (!error.isEmpty()) {
+            return error;
         }
 
         // compute
         String datasetFilename = datasetService.getDatasetFilename(taskDescription.getDatasetUri());
-        stdout = dockerComputationService.dockerRunAndGetLogs(taskDescription, datasetFilename);
+        message = dockerComputationService.dockerRunAndGetLogs(taskDescription, datasetFilename);
 
-        if (stdout.isEmpty()) {
+        if (message.isEmpty()) {
             customFeignClient.updateReplicateStatus(chainTaskId, COMPUTE_FAILED);
-            stdout = "Failed to start computation";
-            log.error(stdout + " [chainTaskId:{}]", chainTaskId);
-            return stdout;
+            message = "Failed to start computation";
+            log.error(message + " [chainTaskId:{}]", chainTaskId);
+            return message;
         }
 
         customFeignClient.updateReplicateStatus(chainTaskId, COMPUTED);
-        return stdout;
+        return message;
+    }
+
+    private String tryToDownloadApp(TaskDescription taskDescription) {
+        String chainTaskId = taskDescription.getChainTaskId();
+
+        String error = checkContributionAbility(chainTaskId);
+        if (!error.isEmpty()) {
+            return error;
+        }
+
+        // pull app
+        customFeignClient.updateReplicateStatus(chainTaskId, APP_DOWNLOADING);
+        boolean isAppDownloaded = dockerComputationService.dockerPull(chainTaskId, taskDescription.getAppUri());
+        if (!isAppDownloaded) {
+            customFeignClient.updateReplicateStatus(chainTaskId, APP_DOWNLOAD_FAILED);
+            String errorMessage = "Failed to pull application image, URI:" + taskDescription.getAppUri();
+            log.error(errorMessage + " [chainTaskId:{}]", chainTaskId);
+            return errorMessage;
+        }
+
+        return "";
+    }
+
+    private String tryToDownloadData(TaskDescription taskDescription) {
```

Comment:
getDataDownloadStatus(..)

Reply:
You are right a big refactoring is needed for this entire method.